File: api/src/main/java/com/ruby/rubia_server/core/service/MessageEnhancementAuditService.java
package com.ruby.rubia_server.core.service;

import com.ruby.rubia_server.core.entity.*;
import com.ruby.rubia_server.core.repository.MessageEnhancementAuditRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
@Slf4j
@Transactional
public class MessageEnhancementAuditService {

    private final MessageEnhancementAuditRepository auditRepository;

    /**
     * Registra uma melhoria de mensagem bem-sucedida
     */
    public MessageEnhancementAudit recordSuccessfulEnhancement(
            Company company,
            User user,
            AIAgent aiAgent,
            String originalMessage,
            String enhancedMessage,
            UUID conversationId,
            Integer tokensConsumed,
            Long responseTimeMs,
            String userAgent,
            String ipAddress,
            String openaiSystemMessage,
            String openaiUserMessage,
            String openaiFullPayload) {

        log.debug("Recording successful message enhancement for user: {} with agent: {}", 
                 user.getId(), aiAgent.getName());

        MessageEnhancementAudit audit = MessageEnhancementAudit.builder()
                .company(company)
                .user(user)
                .aiAgent(aiAgent)
                .conversationId(conversationId)
                .originalMessage(originalMessage)
                .enhancedMessage(enhancedMessage)
                .temperamentUsed(aiAgent.getTemperament())
                .aiModelUsed(aiAgent.getAiModel().getName())
                .temperatureUsed(aiAgent.getTemperature().doubleValue())
                .maxTokensUsed(aiAgent.getMaxResponseLength())
                .tokensConsumed(tokensConsumed)
                .responseTimeMs(responseTimeMs)
                .success(true)
                .userAgent(userAgent)
                .ipAddress(ipAddress)
                .openaiSystemMessage(openaiSystemMessage)
                .openaiUserMessage(openaiUserMessage)
                .openaiFullPayload(openaiFullPayload)
                .build();

        MessageEnhancementAudit saved = auditRepository.save(audit);
        log.info("Message enhancement audit recorded successfully with ID: {}", saved.getId());
        
        return saved;
    }

    /**
     * Registra uma melhoria de mensagem que falhou
     */
    public MessageEnhancementAudit recordFailedEnhancement(
            Company company,
            User user,
            AIAgent aiAgent,
            String originalMessage,
            String errorMessage,
            UUID conversationId,
            Long responseTimeMs,
            String userAgent,
            String ipAddress,
            String openaiSystemMessage,
            String openaiUserMessage,
            String openaiFullPayload) {

        log.debug("Recording failed message enhancement for user: {} with agent: {}", 
                 user.getId(), aiAgent.getName());

        MessageEnhancementAudit audit = MessageEnhancementAudit.builder()
                .company(company)
                .user(user)
                .aiAgent(aiAgent)
                .conversationId(conversationId)
                .originalMessage(originalMessage)
                .enhancedMessage(null) // Não há mensagem melhorada em caso de falha
                .temperamentUsed(aiAgent.getTemperament())
                .aiModelUsed(aiAgent.getAiModel().getName())
                .temperatureUsed(aiAgent.getTemperature().doubleValue())
                .maxTokensUsed(aiAgent.getMaxResponseLength())
                .tokensConsumed(null) // Não houve consumo de tokens em falha
                .responseTimeMs(responseTimeMs)
                .success(false)
                .errorMessage(errorMessage)
                .userAgent(userAgent)
                .ipAddress(ipAddress)
                .openaiSystemMessage(openaiSystemMessage)
                .openaiUserMessage(openaiUserMessage)
                .openaiFullPayload(openaiFullPayload)
                .build();

        MessageEnhancementAudit saved = auditRepository.save(audit);
        log.warn("Message enhancement failure recorded with ID: {} - Error: {}", saved.getId(), errorMessage);
        
        return saved;
    }

    /**
     * Busca auditorias por empresa com paginação
     */
    @Transactional(readOnly = true)
    public Page<MessageEnhancementAudit> getAuditsByCompany(UUID companyId, Pageable pageable) {
        return auditRepository.findByCompanyId(companyId, pageable);
    }

    /**
     * Busca auditorias por usuário com paginação
     */
    @Transactional(readOnly = true)
    public Page<MessageEnhancementAudit> getAuditsByUser(UUID userId, Pageable pageable) {
        return auditRepository.findByUserId(userId, pageable);
    }

    /**
     * Busca auditorias por agente IA com paginação
     */
    @Transactional(readOnly = true)
    public Page<MessageEnhancementAudit> getAuditsByAiAgent(UUID aiAgentId, Pageable pageable) {
        return auditRepository.findByAiAgentId(aiAgentId, pageable);
    }

    /**
     * Busca auditorias por conversa
     */
    @Transactional(readOnly = true)
    public List<MessageEnhancementAudit> getAuditsByConversation(UUID conversationId) {
        return auditRepository.findByConversationId(conversationId);
    }

    /**
     * Estatísticas de melhorias por empresa
     */
    @Transactional(readOnly = true)
    public EnhancementStats getCompanyStats(UUID companyId) {
        long successful = auditRepository.countSuccessfulEnhancementsByCompany(companyId);
        long failed = auditRepository.countFailedEnhancementsByCompany(companyId);
        long totalTokens = auditRepository.sumTokensConsumedByCompany(companyId);
        Double avgResponseTime = auditRepository.averageResponseTimeByCompany(companyId);

        return EnhancementStats.builder()
                .successfulEnhancements(successful)
                .failedEnhancements(failed)
                .totalEnhancements(successful + failed)
                .totalTokensConsumed(totalTokens)
                .averageResponseTimeMs(avgResponseTime != null ? avgResponseTime : 0.0)
                .successRate(successful + failed > 0 ? (double) successful / (successful + failed) * 100 : 0.0)
                .build();
    }

    /**
     * Busca auditorias por período
     */
    @Transactional(readOnly = true)
    public Page<MessageEnhancementAudit> getAuditsByDateRange(
            UUID companyId, 
            LocalDateTime startDate, 
            LocalDateTime endDate, 
            Pageable pageable) {
        return auditRepository.findByCompanyIdAndDateRange(companyId, startDate, endDate, pageable);
    }

    /**
     * Busca auditorias por temperamento
     */
    @Transactional(readOnly = true)
    public Page<MessageEnhancementAudit> getAuditsByTemperament(
            UUID companyId, 
            String temperament, 
            Pageable pageable) {
        return auditRepository.findByCompanyIdAndTemperament(companyId, temperament, pageable);
    }

    /**
     * Busca auditorias por modelo de IA
     */
    @Transactional(readOnly = true)
    public Page<MessageEnhancementAudit> getAuditsByAiModel(
            UUID companyId, 
            String model, 
            Pageable pageable) {
        return auditRepository.findByCompanyIdAndAiModel(companyId, model, pageable);
    }

    /**
     * Classe para estatísticas de enhancement
     */
    @lombok.Data
    @lombok.Builder
    public static class EnhancementStats {
        private long successfulEnhancements;
        private long failedEnhancements;
        private long totalEnhancements;
        private long totalTokensConsumed;
        private double averageResponseTimeMs;
        private double successRate; // Porcentagem
    }
}

File: api-dotnet/Rubia.Server/Services/MessageEnhancementAuditService.cs
using Microsoft.EntityFrameworkCore;
using Rubia.Server.Data;
using Rubia.Server.Entities;

namespace Rubia.Server.Services;

public class MessageEnhancementAuditService
{
    private readonly RubiaDbContext _context;
    private readonly ILogger<MessageEnhancementAuditService> _logger;

    public MessageEnhancementAuditService(
        RubiaDbContext context,
        ILogger<MessageEnhancementAuditService> logger)
    {
        _context = context;
        _logger = logger;
    }

    public async Task<MessageEnhancementAudit> RecordSuccessfulEnhancementAsync(
        Company company,
        User user,
        AIAgent aiAgent,
        string originalMessage,
        string enhancedMessage,
        Guid? conversationId,
        int? tokensConsumed,
        long? responseTimeMs,
        string? userAgent,
        string? ipAddress,
        string? openaiSystemMessage,
        string? openaiUserMessage,
        string? openaiFullPayload)
    {
        _logger.LogDebug("Recording successful message enhancement for user: {UserId} with agent: {AgentName}",
            user?.Id, aiAgent.Name);

        var audit = new MessageEnhancementAudit
        {
            Company = company,
            CompanyId = company.Id,
            User = user,
            UserId = user.Id,
            AiAgent = aiAgent,
            AiAgentId = aiAgent.Id,
            ConversationId = conversationId,
            OriginalMessage = originalMessage,
            EnhancedMessage = enhancedMessage,
            TemperamentUsed = aiAgent.Temperament ?? string.Empty,
            AiModelUsed = aiAgent.AiModel?.Name ?? "unknown",
            TemperatureUsed = (double?)aiAgent.Temperature ?? 0.7,
            MaxTokensUsed = aiAgent.MaxResponseLength,
            TokensConsumed = tokensConsumed,
            ResponseTimeMs = responseTimeMs,
            Success = true,
            UserAgent = userAgent,
            IpAddress = ipAddress,
            OpenaiSystemMessage = openaiSystemMessage,
            OpenaiUserMessage = openaiUserMessage,
            OpenaiFullPayload = openaiFullPayload
        };

        _context.MessageEnhancementAudits.Add(audit);
        await _context.SaveChangesAsync();

        _logger.LogInformation("Message enhancement audit recorded successfully with ID: {AuditId}", audit.Id);
        return audit;
    }

    public async Task<MessageEnhancementAudit> RecordFailedEnhancementAsync(
        Company company,
        User user,
        AIAgent aiAgent,
        string originalMessage,
        string errorMessage,
        Guid? conversationId,
        long? responseTimeMs,
        string? userAgent,
        string? ipAddress,
        string? openaiSystemMessage,
        string? openaiUserMessage,
        string? openaiFullPayload)
    {
        _logger.LogDebug("Recording failed message enhancement for user: {UserId} with agent: {AgentName}",
            user?.Id, aiAgent.Name);

        var audit = new MessageEnhancementAudit
        {
            Company = company,
            CompanyId = company.Id,
            User = user,
            UserId = user.Id,
            AiAgent = aiAgent,
            AiAgentId = aiAgent.Id,
            ConversationId = conversationId,
            OriginalMessage = originalMessage,
            EnhancedMessage = null,
            TemperamentUsed = aiAgent.Temperament ?? string.Empty,
            AiModelUsed = aiAgent.AiModel?.Name ?? "unknown",
            TemperatureUsed = (double?)aiAgent.Temperature ?? 0.7,
            MaxTokensUsed = aiAgent.MaxResponseLength,
            TokensConsumed = null,
            ResponseTimeMs = responseTimeMs,
            Success = false,
            ErrorMessage = errorMessage,
            UserAgent = userAgent,
            IpAddress = ipAddress,
            OpenaiSystemMessage = openaiSystemMessage,
            OpenaiUserMessage = openaiUserMessage,
            OpenaiFullPayload = openaiFullPayload
        };

        _context.MessageEnhancementAudits.Add(audit);
        await _context.SaveChangesAsync();

        _logger.LogWarning("Message enhancement failure recorded with ID: {AuditId} - Error: {Error}",
            audit.Id, errorMessage);
        return audit;
    }

    public async Task<(List<MessageEnhancementAudit> Items, int TotalCount)> GetAuditsByCompanyAsync(
        Guid companyId, int page, int pageSize)
    {
        var query = _context.MessageEnhancementAudits
            .Include(a => a.User)
            .Include(a => a.AiAgent)
            .Where(a => a.CompanyId == companyId)
            .OrderByDescending(a => a.CreatedAt);

        var totalCount = await query.CountAsync();
        var items = await query
            .Skip(page * pageSize)
            .Take(pageSize)
            .ToListAsync();

        return (items, totalCount);
    }

    public async Task<(List<MessageEnhancementAudit> Items, int TotalCount)> GetAuditsByUserAsync(
        Guid userId, int page, int pageSize)
    {
        var query = _context.MessageEnhancementAudits
            .Include(a => a.User)
            .Include(a => a.AiAgent)
            .Where(a => a.UserId == userId)
            .OrderByDescending(a => a.CreatedAt);

        var totalCount = await query.CountAsync();
        var items = await query
            .Skip(page * pageSize)
            .Take(pageSize)
            .ToListAsync();

        return (items, totalCount);
    }

    public async Task<(List<MessageEnhancementAudit> Items, int TotalCount)> GetAuditsByAiAgentAsync(
        Guid aiAgentId, int page, int pageSize)
    {
        var query = _context.MessageEnhancementAudits
            .Include(a => a.User)
            .Include(a => a.AiAgent)
            .Where(a => a.AiAgentId == aiAgentId)
            .OrderByDescending(a => a.CreatedAt);

        var totalCount = await query.CountAsync();
        var items = await query
            .Skip(page * pageSize)
            .Take(pageSize)
            .ToListAsync();

        return (items, totalCount);
    }

    public async Task<List<MessageEnhancementAudit>> GetAuditsByConversationAsync(Guid conversationId)
    {
        return await _context.MessageEnhancementAudits
            .Include(a => a.User)
            .Include(a => a.AiAgent)
            .Where(a => a.ConversationId == conversationId)
            .OrderByDescending(a => a.CreatedAt)
            .ToListAsync();
    }

    public async Task<EnhancementStats> GetCompanyStatsAsync(Guid companyId)
    {
        var audits = await _context.MessageEnhancementAudits
            .Where(a => a.CompanyId == companyId)
            .ToListAsync();

        var successful = audits.Count(a => a.Success);
        var failed = audits.Count(a => !a.Success);
        var totalTokens = audits.Where(a => a.TokensConsumed.HasValue).Sum(a => a.TokensConsumed.Value);
        var avgResponseTime = audits.Where(a => a.ResponseTimeMs.HasValue)
            .Average(a => (double?)a.ResponseTimeMs) ?? 0.0;

        return new EnhancementStats
        {
            SuccessfulEnhancements = successful,
            FailedEnhancements = failed,
            TotalEnhancements = successful + failed,
            TotalTokensConsumed = totalTokens,
            AverageResponseTimeMs = avgResponseTime,
            SuccessRate = successful + failed > 0 ? (double)successful / (successful + failed) * 100 : 0.0
        };
    }

    public async Task<(List<MessageEnhancementAudit> Items, int TotalCount)> GetAuditsByDateRangeAsync(
        Guid companyId, DateTime startDate, DateTime endDate, int page, int pageSize)
    {
        var query = _context.MessageEnhancementAudits
            .Include(a => a.User)
            .Include(a => a.AiAgent)
            .Where(a => a.CompanyId == companyId &&
                       a.CreatedAt >= startDate &&
                       a.CreatedAt <= endDate)
            .OrderByDescending(a => a.CreatedAt);

        var totalCount = await query.CountAsync();
        var items = await query
            .Skip(page * pageSize)
            .Take(pageSize)
            .ToListAsync();

        return (items, totalCount);
    }

    public async Task<(List<MessageEnhancementAudit> Items, int TotalCount)> GetAuditsByTemperamentAsync(
        Guid companyId, string temperament, int page, int pageSize)
    {
        var query = _context.MessageEnhancementAudits
            .Include(a => a.User)
            .Include(a => a.AiAgent)
            .Where(a => a.CompanyId == companyId &&
                       a.TemperamentUsed == temperament)
            .OrderByDescending(a => a.CreatedAt);

        var totalCount = await query.CountAsync();
        var items = await query
            .Skip(page * pageSize)
            .Take(pageSize)
            .ToListAsync();

        return (items, totalCount);
    }

    public async Task<(List<MessageEnhancementAudit> Items, int TotalCount)> GetAuditsByAiModelAsync(
        Guid companyId, string model, int page, int pageSize)
    {
        var query = _context.MessageEnhancementAudits
            .Include(a => a.User)
            .Include(a => a.AiAgent)
            .Where(a => a.CompanyId == companyId &&
                       a.AiModelUsed == model)
            .OrderByDescending(a => a.CreatedAt);

        var totalCount = await query.CountAsync();
        var items = await query
            .Skip(page * pageSize)
            .Take(pageSize)
            .ToListAsync();

        return (items, totalCount);
    }
}

public class EnhancementStats
{
    public long SuccessfulEnhancements { get; set; }
    public long FailedEnhancements { get; set; }
    public long TotalEnhancements { get; set; }
    public long TotalTokensConsumed { get; set; }
    public double AverageResponseTimeMs { get; set; }
    public double SuccessRate { get; set; }
}